core/src/main/java/com/oath/oak/Chunk.java

```diff
@@ -237,68 +292,111 @@ ByteBuffer readMaxKey() {
     /**
      * gets the field of specified offset for given item in entry array
      */
-    private int getEntryField(int item, int offset) {
+    private int getEntryFieldInt(int item, OFFSET offset) {
         switch (offset) {
-            case OFFSET_KEY_LENGTH:
+            case KEY_LENGTH:
                 // return two low bytes of the key length index int
-                return (entries[item + OFFSET_KEY_LENGTH] & KEY_LENGTH_MASK);
-            case OFFSET_KEY_BLOCK:
+                return (entries[item + offset.value] & KEY_LENGTH_MASK);
+            case KEY_BLOCK:
                 // offset must be OFFSET_KEY_BLOCK, return 2 high bytes of the int inside key length
                 // right-shift force, fill empty with zeroes
-                return (entries[item + OFFSET_KEY_LENGTH] >>> KEY_BLOCK_SHIFT);
+                return (entries[item + offset.value] >>> KEY_BLOCK_SHIFT);
+            case VALUE_LENGTH:
+                return (entries[item + offset.value] & VALUE_LENGTH_MASK);
+            case VALUE_BLOCK:
+                return (entries[item + offset.value] >>> VALUE_BLOCK_SHIFT);
             default:
-                return entries[item + offset];
+                return entries[item + offset.value];
         }
     }
 
+    // Atomically reads two integers of the entries array.
+    // Should be used with OFFSET.VALUE_REFERENCE and OFFSET.KEY_REFERENCE
+    private long getEntryFieldLong(int item, OFFSET offset) {
+        long arrayOffset = Unsafe.ARRAY_INT_BASE_OFFSET + (item + offset.value) * Unsafe.ARRAY_INT_INDEX_SCALE;
+        assert arrayOffset % 8 == 0;
+        return unsafe.getLongVolatile(entries, arrayOffset);
+    }
+
     /**
      * * sets the field of specified offset to 'value' for given item in entry array
+     * NOT ATOMIC!
```

Comment:
I think that an assignment of a machine word **is** atomic, but more likely doesn't propagate to other threads without additional memory fence...